lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource'
+require 'kubernetes-deploy/kubeclient_builder'
+require "active_record"
+require "jsonpath"
+
+module KubernetesDeploy
+  class DiscoverableResource < KubernetesResource
+    extend KubernetesDeploy::KubeclientBuilder
+
+    class << self
+      attr_accessor :version, :group, :type
+      attr_reader :timeout
+    end
+
+    STATUS_FIELD_ANNOTATION = 'kubernetes-deploy.shopify.io/status-field'
+    STATUS_SUCCESS_ANNOTATION = 'kubernetes-deploy.shopify.io/status-success'
+    TIMEOUT_ANNOTATION = 'kubernetes-deploy.shopify.io/timeout'
+    PREDEPLOY_ANNOTATION = 'kubernetes-deploy.shopify.io/predeploy'
+    PRUNABLE_ANNOTATION = 'kubernetes-deploy.shopify.io/prunable'
```

Comment:
There are so many annotations here that I'm wondering if we should instead have a single one that contains JSON.